src/python_testing/TC_TSTAT_4_2.py

```diff
@@ -377,67 +474,106 @@ async def test_TC_TSTAT_4_2(self):
+                presetType.presetScenario, 0) < presetType.numberOfPresets)
 
-            # Write to the presets attribute after setting a name for preset with handle (b'\x01') that doesn't support names
-            test_presets = copy.deepcopy(new_presets_with_handle)
-            test_presets[0].name = "Occupied"
+            test_presets = copy.deepcopy(current_presets)
+            presets_without_name_support = list(preset for preset in test_presets if preset.presetScenario in availableScenarios)
 
-            await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ConstraintError)
+            if len(presets_without_name_support) == 0 and len(availableScenarios) > 0:
+                new_preset = cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=availableScenarios[0],
+                                                          coolingSetpoint=coolSetpoint, heatingSetpoint=heatSetpoint, builtIn=True)
+                test_presets.append(new_preset)
+                presets_without_name_support = [new_preset]
 
-            # Clear state for next test.
-            await self.send_atomic_request_rollback_command()
+            if len(availableScenarios) > 0:
+
+                # Write to the presets attribute after setting a name for preset with handle (b'\x01') that doesn't support names
+                presets_without_name_support[0].name = "Name"
+
+                # Send the AtomicRequest begin command
+                await self.send_atomic_request_begin_command()
+
+                await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ConstraintError)
+
+                # Clear state for next test.
+                await self.send_atomic_request_rollback_command()
+            else:
+                logger.info(
+                    "Couldn't run test step 12 since there was no available preset scenario without name support")
 
         self.step("13")
         if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
 
-            # Send the AtomicRequest begin command
-            await self.send_atomic_request_begin_command()
+            availableScenario = self.get_available_scenario(presetTypes=presetTypes, presetScenarioCounts=presetScenarioCounts)
 
-            # Write to the presets attribute with a new valid preset added
-            test_presets = copy.deepcopy(new_presets_with_handle)
-            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kWake,
-                                name="Wake", coolingSetpoint=2800, heatingSetpoint=1800, builtIn=False))
+            if len(current_presets) > 0 and len(current_presets) < numberOfPresetsSupported and availableScenario is not None:
```

Comment:
Why do we care whether `len(current_presets) > 0`?

Reply:
We don't; fixed.